### PR TITLE
Select multiple raw material items for Sub Contract Stock Entries

### PR DESCRIPTION
**Pull request by PawanMeh**
Testing Video

subcontract testing

1. Allow selection of multiple items from Sub Contract PO based on BOM.
2. Group raw material stock entries by BOM Finished Goods and create the stock entries.

### CODE REVIEW

**Comment by PawanMeh** [comment]
Checked Travis errors, do not seem to be related to this change. Assigning for review.

File: erpnext/buying/doctype/purchase_order/purchase_order.py [hunk 1]
```diff
@@ -415,6 +415,54 @@ def make_stock_entry(purchase_order, item_code):
 	stock_entry.get_items()
 	return stock_entry.as_dict()
```
Comment (rohitwaghchaure) [inline erpnext/buying/doctype/purchase_order/purchase_order.py]: There is two methods make_stock_entry, make_rm_stock_entry for the same functionality, can you commonify it?

File: erpnext/buying/doctype/purchase_order/purchase_order.py [hunk 2]
```diff
@@ -415,6 +415,54 @@ def make_stock_entry(purchase_order, item_code):
...
+			for rm_item_data in rm_items_list:
+				if rm_item_data["item_code"] == item_code:
+					items_dict = {rm_item_data["rm_item_code"]:
+					{"item_name":rm_item_data["item_name"],
+					"description":item_wh.get(rm_item_data["rm_item_code"]),
+					'qty':rm_item_data["qty"],
+					'from_warehouse':rm_item_data["warehouse"],
+					'stock_uom':rm_item_data["stock_uom"]}}
+					stock_entry.add_to_stock_entry_detail(items_dict)
+			stock_entry.save()
```
Comment (rohitwaghchaure) [inline erpnext/buying/doctype/purchase_order/purchase_order.py]: Why save separate stock entry for each item, you can make single stock entry against selected items and no need to save. On click of transfer button open single unsaved stock entry with the selected items.

File: erpnext/buying/doctype/purchase_order/purchase_order.py (hunk 1 above)
Comment (nabinhait) [reply erpnext/buying/doctype/purchase_order/purchase_order.py]: I think `make_stock_entry` does not required anymore.

File: erpnext/stock/doctype/stock_entry/stock_entry.py [hunk 3]
```diff
@@ -64,6 +64,14 @@ def validate(self):
...
 
 	def on_submit(self):
+		if self.purpose == "Subcontract":
+			for d in self.items:
+				if not d.t_warehouse:
+					if self.to_warehouse:
+						d.t_warehouse = self.to_warehouse
+					else:
+						frappe.throw(_("Target warehouse is mandatory for row {0}").format(d.idx))
+
```
Comment (rohitwaghchaure) [inline erpnext/stock/doctype/stock_entry/stock_entry.py]: No need to make these changes

File: erpnext/stock/doctype/stock_entry/stock_entry.py [hunk 4]
```diff
@@ -157,7 +165,9 @@ def validate_warehouse(self):
 				if self.to_warehouse:
 					d.t_warehouse = self.to_warehouse
 				else:
-					frappe.throw(_("Target warehouse is mandatory for row {0}").format(d.idx))
+					#move validation for sub contract to submit
+					if self.purpose != "Subcontract":
+						frappe.throw(_("Target warehouse is mandatory for row {0}").format(d.idx))
```
Comment (rohitwaghchaure) [inline erpnext/stock/doctype/stock_entry/stock_entry.py]: No need to make these changes

File: erpnext/buying/doctype/purchase_order/purchase_order.py (hunk 1 above)
Comment (PawanMeh) [reply erpnext/buying/doctype/purchase_order/purchase_order.py]: If we don't want to maintain old way of stock entry creation for one finished good, we can just change if condition in JS to call the new modal in all cases and then remove the old method.

Comment (rohitwaghchaure) [reply erpnext/buying/doctype/purchase_order/purchase_order.py]: Have checked the code and found that BOM No is used for reference only. Can we create the single stock entry against all the items without BOM No?

Comment (PawanMeh) [reply erpnext/buying/doctype/purchase_order/purchase_order.py]: In that case, what should we set in stock entry (finished good completed quantity, should it be sum of all finished goods item completed quantities? or should it be zero?)
stock_entry.fg_completed_qty = po_item.qty

**Comment by PawanMeh** [comment]
Made changes to make single stock entry instead of multiple & update bom no at stock entry detail level

File: erpnext/buying/doctype/purchase_order/purchase_order.py (hunk 1 above)
Comment (rohitwaghchaure) [reply erpnext/buying/doctype/purchase_order/purchase_order.py]: Can you remove the old method make_stock_entry, we can use method make_rm_stock_entry if one item is there.

**Comment by PawanMeh** [comment]
Made changes, travis error does not seem to be related to this. Manual testing video attached

subcontract stock entry

**Comment by PawanMeh** [comment]
Removed old stock entry method and rewrote test case accordingly.

**Review by rohitwaghchaure — APPROVED** [review]

**Pull request merged**